src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -207,66 +207,69 @@ private ProjectId GetActiveContextProjectId(IVsHierarchy hierarchy, IEnumerable<
+                if (ErrorHandler.Succeeded(hierarchy.GetProperty(VSConstants.VSITEMID_ROOT, (int)__VSHPROPID8.VSHPROPID_ActiveIntellisenseProjectContext, out object contextProjectNameObject)))
+                {
+                    EnsureSubscribedToHierarchyPropertyChanges(hierarchy);
 
-                    // This is some intermediate hierarchy which we need to direct us somewhere else. At this point, we need to add an event sink to be aware if the redirection
-                    // ever changes.
-                    if (!_hierarchyEventSinks.ContainsKey(hierarchy))
+                    if (contextProjectNameObject is string contextProjectName)
                     {
-                        var eventSink = new HierarchyEventSink(intermediateHierarchy, this);
-                        if (eventSink.TryAdviseHierarchy())
+                        var project = _workspace.GetProjectWithHierarchyAndName(hierarchy, contextProjectName);
+
+                        if (project != null && projectIds.Contains(project.Id))
                         {
-                            _hierarchyEventSinks.Add(intermediateHierarchy, eventSink);
+                            return project.Id;
                         }
                     }
+                }
 
-                    // If this is a shared hierarchy, we can possibly ask it for it's context
-                    var contextHierarchy = intermediateHierarchy.GetActiveProjectContext();
-                    if (contextHierarchy != null)
-                    {
-                        intermediateHierarchy = contextHierarchy;
-                        continue;
-                    }
+                // At this point, we should hopefully have only one project that maches by hierarchy. If there's multiple, at this point we can't figure anything
+                // out better.
+                var matchingProjectId = projectIds.FirstOrDefault(id => projectToHierarchyMap.GetValueOrDefault(id) == hierarchy);
 
-                    if (ErrorHandler.Succeeded(intermediateHierarchy.GetProperty(VSConstants.VSITEMID_ROOT, (int)__VSHPROPID8.VSHPROPID_ActiveIntellisenseProjectContext, out object contextProjectNameObject)))
-                    {
-                        if (contextProjectNameObject is string contextProjectName)
-                        {
-                            var contextProject = solution.Projects.FirstOrDefault(p => p.Name == contextProjectName);
-                            if (contextProject != null)
-                            {
-                                return contextProject.Id;
-                            }
-                        }
-                    }
+                if (matchingProjectId != null)
+                {
+                    return matchingProjectId;
                 }
 
                 // If we had some trouble finding the project, we'll just pick one arbitrarily
                 return projectIds.First();
```

Comment:
❔ Do we not want to use Miscellaneous Files for this?

Reply:
Not really. projectIds in this case is a list of projects that already contains this file, presumably they have the right references, or at least much better references than Miscellaneous Files would provide.

This is really the "last resort" call; I should add a comment clarifying that we really don't ever want to end up here, but if anything goes wrong with another project system, we can't entirely blow up.